### PR TITLE
26 integrate modular anns

### LINKED ISSUE #26: Integrate Modular ANNs

### PR DESCRIPTION
**Pull request by PatrickHenkel1**
# Pull Request

## Description
Adds modular ANNs to physXAI. Modular ANNs allow more flexibility to create new models, e.g. subsystems with linear and non-linear inputs

## Type of Change
- [ ] Bug fix (non-breaking change fixing an issue)
- [X] New feature (non-breaking change adding functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (code change that neither fixes a bug nor adds a feature)
- [ ] Documentation update

### Testing
- [ ] Unit tests have been created/updated for new/modified functionality
- [x] CI/CD pipeline passes all tests (pytest, coverage, executables, installation)

### Examples
- [x] Add examples for new features and functionality

### Compatibility
- [X] Changes are backward compatible OR deprecation warnings added
- [X] No breaking changes to public APIs
- [X] New dependencies added to `pyproject.toml` (required in `dependencies` or optional in `[project.optional-dependencies]`)

### Documentation
- [x] Docstrings added/updated for new/modified public methods (Google style)
- [x] Type hints added for new functions/methods

### GitHub Copilot Review
- [X] Request Copilot review via GitHub UI (add 'copilot' as a reviewer)

### CODE REVIEW

File: physXAI/models/modular/modular_ann.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+            if isinstance(self.model, LinearRegressionModel):
+                lr = ModularLinear(inputs=self.inputs, name=self.name + "_linear").construct(input_layer, td)
+                l = lr(keras.layers.Concatenate()(inps))
+            else:
+                l = self.model.generate_model(td=td)(keras.layers.Concatenate()(inps))
+            if self.rescale_output:
+                l = keras.layers.Rescaling(scale=self.nominal_sigma, offset=self.nominal_mean)(l) 
+            ModularExpression.models[self.name] = l
+            return l    
+
```
Comment (SiAndRo2002) [inline physXAI/models/modular/modular_ann.py]: get_config() method useful?
Useful to store nominal_range etc.?

File: physXAI/models/modular/modular_ann.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        self.inputs = [inp if isinstance(inp, ModularExpression) else inp.input() for inp in inputs]
+
+        if nominal_range is None:
+            self.rescale_output = False
+        elif nominal_range is not None and len(nominal_range) != 2:
+            raise ValueError(f"Modular Model: nominal_range must be a tuple of (min, max), but was {nominal_range}")
+        else:
+            self.rescale_output = True
+            self.nominal_mean = (nominal_range[1] + nominal_range[0]) / 2.0
+            self.nominal_sigma = (nominal_range[1] - nominal_range[0]) / 4.0  # Assuming 4 sigma covers the range
```
Comment (SiAndRo2002) [inline physXAI/models/modular/modular_ann.py]: This structure is used in many Modular classes and could be moved into a separate function

File: physXAI/models/modular/modular_ann.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+        if name is None:
+            name = f"ModularNormalization_{ModularNormalization.i}"
+            ModularNormalization.i += 1
+        super().__init__(name)
+        self.inputs = input
+
+    def construct(self, input_layer: keras.layers.Input, td: TrainingDataGeneric) -> keras.layers.Layer:
+        inp = self.inputs.construct(input_layer, td)
+        normalization = keras.layers.BatchNormalization()
+        l = normalization(inp)
```
Comment (SiAndRo2002) [inline physXAI/models/modular/modular_ann.py]: Is it intended that ModularExpression.models[self.name] = l is missing here?

Comment (PatrickHenkel1) [reply physXAI/models/modular/modular_ann.py]: In my opionion it is not really useful to use an existing BatchNormalization twice. Therefore, for normalization there is no model save

File: physXAI/models/modular/modular_ann.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+    def construct(self, input_layer: keras.layers.Input, td: TrainingDataGeneric) -> keras.layers.Layer:
+        if self.name in ModularExpression.models.keys():
+            return ModularExpression.models[self.name]
+        else:
+            inps = list()
+            for x in self.inputs:
+                y = x.construct(input_layer, td)
+                inps.append(y)
+            self.model.model_config['n_features'] = len(inps)
```
[the remark this answers is not in the document]
Comment (PatrickHenkel1) [reply physXAI/models/modular/modular_ann.py]: Not sure if I understand the question. The if statement is returning a existing model and if this does not exist, a new model is created

File: physXAI/models/modular/modular_ann.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        if name is None:
+            name = f"ModularModel_{ModularModel.i}"
+            ModularModel.i += 1
+
+        super().__init__(name)
+        self.model = model
+        self.model.model_config.update({
+            "normalize": False,
+            "rescale_output": False
+        })
```
[the remark this answers is not in the document]
Comment (PatrickHenkel1) [reply physXAI/models/modular/modular_ann.py]: As the individual features are normalized, normalization of a hidden ANN model is not really necessary. With rescale__output the model output is scaled to the final output, which is also not really useful for hidden ANNs. Still, a user warning might be a good idea

File: unittests/modular/test_modular.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    output = "x4"
+    
+    features = list()
+    for inp in inputs:
+        features.append(Feature(inp))
+
+    prep = PreprocessingSingleStep(inputs=inputs, output=output, random_state=random_seed)
+    td = prep.pipeline(training_data_path)
+
+    # TODO: Flatten, BatchNorm, Cropping1D, Reshape, RBF
```
Comment (SiAndRo2002) [inline unittests/modular/test_modular.py]: Complete TODO

**Review by SiAndRo2002 — CHANGES_REQUESTED** [review]
See comments
